### ISSUE TITLE
DDP grads dont have parity with local training when grads are undefined

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#58522]
## 🐛 Bug
The following test fails for DDP:

```
@unittest.skipIf(BACKEND != 'nccl' and BACKEND != 'gloo',
                         "Only Nccl & Gloo backend support DistributedDataParallel")
        @skip_if_lt_x_gpu(2)
        def test_undefined_grad_parity_unused_parameters(self):
            # TODO: enable this for general training use cases.
            x = torch.ones(1, 2).to(self.rank)
            net = Net().to(self.rank)
            local_net = copy.deepcopy(net)
            net = torch.nn.parallel.DistributedDataParallel(
                net,
                device_ids=[self.rank],
                find_unused_parameters=False,
            )
            out = net(x).sum()
            local_out = local_net(x).sum()
            # Simulates undefined gradients.
            torch._C._functions.UndefinedGrad()(out).backward()
            torch._C._functions.UndefinedGrad()(local_out).backward()
            for (dist_param_name, dist_param), (local_param_name, local_param) in zip(net.named_parameters(), local_net.named_parameters()):
                dist_grad = dist_param.grad
                local_grad = local_param.grad
                self.assertEqual(dist_grad, local_grad, f"DDP param {dist_param_name} grad {dist_grad} does not match local param {local_param_name} with grad {local_grad}")
```

The main reason is that if an autograd hook fires in the reducer, we will consider that param as having a grad and allreduce it. If the grad is not defined, we zero the part of the corresponding bucket and proceed, so we end up writing zero to the grad field.

Historically, the main use case of `find_unused_parameters=True` has been how we get undefined gradients, DDP works well in this case by checking if the param hasn't been used across all ranks, and if so, avoids allreducing it. However, in the general case where the param's grad is `None` it does not work as seen above.

I'm not actually sure if there are any practical scenarios where the grad can be `None` but the parameter is still used, however. This could potentially happen if the user has a custom autograd function following the DDP model, that returns `None` for gradients depending on some custom logic. There also have not been any user reports of this issue, so this is probably a low-pri fix for now.

**Comment by rohan-varma** [comment]
Instead of using the `UndefinedGrad`, a torch.autograd.Function as follows can also simulate the issue:
```
# Simulates undefined gradients.
            from torch.autograd import Function
            class MyFunc(Function):
                @staticmethod
                def forward(ctx, *inputs):
                    return inputs

                @staticmethod
                def backward(ctx, *grad_outputs):
                    # Set all incoming gradients to None
                    outputs = []
                    for grad_output in grad_outputs:
                        outputs.append(None)
                    return (None, *outputs)
```

basically the behavior may not be correct if user has a custom function that can return `None` for some grads.

**Comment by wanchaol** [comment]
Hmm setting all grad as None seems not a common use case. For the case, could we do the none check and not trigger allreduce if the grads are None?

**Comment by rohan-varma** [comment]
> Hmm setting all grad as None seems not a common use case. For the case, could we do the none check and not trigger allreduce if the grads are None?

Agreed all grads are None is not a common use case, although the user could conceivably write a custom autograd function that sets some of the grads to None and currently we would end up all reducing zeros today.

The tricky part of implementing your suggestion is we need the entire gang to agree whether to allreduce a grad or not. If one rank gets None and the others don't, then we should still allreduce, but discovering this requires an additional comm. which we only do for find_unused_params=True setting.
